record/record_test.go

```diff
@@ -312,3 +313,156 @@ func TestStaleWriter(t *testing.T) {
 		t.Fatalf("stale write #1: unexpected error: %v", err)
 	}
 }
+
+func TestBasicReadRecovery(t *testing.T) {
+	records := [][]byte{
+		[]byte(strings.Repeat("a", blockSize-headerSize)),
+		[]byte(strings.Repeat("b", blockSize-headerSize)),
+		[]byte(strings.Repeat("c", blockSize-headerSize)),
+	}
+
+	buf := new(bytes.Buffer)
+	w := NewWriter(buf)
+
+	for i := 0; i < len(records); i++ {
+		wRec, err := w.Next()
+		if err != nil {
+			t.Fatal(err)
+		}
+		if _, err = wRec.Write(records[i]); err != nil {
+			t.Fatal(err)
+		}
+	}
+	w.Close() // Finalize the last record.
+
+	rawBufSlice := buf.Bytes()
+
+	// Corrupt the checksum of the second record in our file.
+	rawBufSlice[blockSize+0] = 0xef
+	rawBufSlice[blockSize+1] = 0xbe
+	rawBufSlice[blockSize+2] = 0xad
+	rawBufSlice[blockSize+3] = 0xde
+
+	// The first record should be read/processed just fine.
+	underlyingReader := bytes.NewReader(rawBufSlice)
+	r := NewReader(underlyingReader)
+	r1, err := r.Next()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	r1Data, err := ioutil.ReadAll(r1)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if !bytes.Equal(r1Data, records[0]) {
+		t.Fatal("Unexpected output in record 1's data.")
+	}
+
+	// This record should present problems since the checksum is wrong.
+	_, err = r.Next()
+	if err == nil {
+		t.Fatal("Expected an error while reading a corrupted record.")
+	}
+
+	if !strings.Contains(err.Error(), "checksum mismatch") {
+		t.Fatalf("Unexpected error returned: %s", err.Error())
```

Comment:
"err.Error()" can be just "err". fmt.Printf and friends like t.Fatalf will automatically look for any String or Error method.